scipy/special/scaled_exp1.c

```diff
@@ -0,0 +1,82 @@
+#include <math.h>
+
+//
+// We use exp1_wrap() from specfun_wrappers.c
+//
+extern double exp1_wrap(double x);
+
+
+//
+// Compute a factor of the exponential integral E1.
+// This is used in scaled_exp1(x) for moderate values of x.
+//
+// The function uses the continued fraction expansion given in equation 5.1.22
+// of Abramowitz & Stegun, "Handbook of Mathematical Functions".
+// For n=1, this is
+//
+//    E1(x) = exp(-x)*C(x)
+//
+// where C(x), expressed in the notation used in A&S, is the continued fraction
+//
+//            1    1    1    2    2    3    3
+//    C(x) = ---  ---  ---  ---  ---  ---  ---  ...
+//           x +  1 +  x +  1 +  x +  1 +  x +
+//
+// Here, we pull a factor of 1/z out of C(x), so
+//
+//    E1(x) = (exp(-x)/x)*F(x)
+//
+// and a bit of algebra gives the continued fraction expansion of F(x) to be
+//
+//            1    1    1    2    2    3    3
+//    F(x) = ---  ---  ---  ---  ---  ---  ---  ...
+//           1 +  x +  1 +  x +  1 +  x +  1 +
+//
+static double
+expint1_factor_cont_frac(double x)
```

Comment:
For the continued fraction evaluation, I think we should use [Lentz's algorithm](https://en.wikipedia.org/wiki/Lentz%27s_algorithm), a numerically stable forwards algorithm that allows one to check convergence at each iteration. This is used for instance in the implementation of exponential integrals from [Numerical Recipes](http://numerical.recipes/book/book.html)  (section 6.3). See section 5.2. for details on Lentz's algorithm. Also see section 5.3. for info on a convergence acceleration method that they also use in their exponential integral implementation which we may or may not want to use.

Reply:
Thanks @steppi.  Just letting you know that there are a few other things that are still lingering at the top of my to-do list, so it might be a while longer before I get back to this.

Reply:
Thanks for the update. I'll probably have time to get to it within a few weeks if you'll be busy through then.

Reply:
Just as a passing by comment, Numerical Recipes come with a BSD incompatible license hence we cannot consult to it for our implementations (https://github.com/scipy/scipy/pull/5716). Thus we are not allowed to look or be inspired from it.

Reply:
> Just as a passing by comment, Numerical Recipes come with a BSD incompatible license hence we cannot consult to it for our implementations (#5716). Thus we are not allowed to look or be inspired from it.

That's a good point, and it's really unfortunate because the explanations in NR tend to be very good, but the code itself is mostly of poor quality. The book itself states

_Copyright does not protect ideas, but only the expression of those ideas in a particular form. In the case of a computer program, the ideas consist of the program's methodology and algorithm, including the necessary sequence of steps adopted by the programmer. The expression of those ideas is the program source code ... If you analyze the ideas contained in a program, and then express those ideas in your own completely different implementation, then that new program implementation belongs to you._

but  the line is completely unclear. I suppose it is better to avoid the book entirely despite it being a great resource for many topics. I can find another reference that has all of the right details, maybe the Boost implementation, and link it here instead, and to be cautious I guess I should sit out implementing Lentz's algorithm here.